pkg/util/http.go

```diff
@@ -296,3 +297,48 @@ func SerializeProtoResponse(w http.ResponseWriter, resp proto.Message, compressi
 	}
 	return nil
 }
+
+// ParseRequestFormWithoutConsumingBody parsed and returns the request parameters (query string and/or request body)
+// from the input http.Request. If the request has a Body, the request's Body is replaces so that it can be consumed again.
+func ParseRequestFormWithoutConsumingBody(r *http.Request) (url.Values, error) {
+	if r.Body == nil {
+		if err := r.ParseForm(); err != nil {
+			return nil, err
+		}
+
+		return r.Form, nil
+	}
+
+	// Close the original body reader. It's going to be replaced later in this function.
+	origBody := r.Body
+	defer func() { _ = origBody.Close() }()
```

Comment:
The `ParseRequestFormWithoutConsumingBody()` logic has been copied from `handler.ServeHTTP()`. However, previously we didn't close the original `Body` (in `handler.ServeHTTP()` we just close the replaced `req.Body`) but I think we actually have to close the original `Body` too, so I've added this.

Reply:
I think you really need to call Close() only if the resource would be leaked otherwise. Like a file descriptor or some pooled memory.  Otherwise GC will just clean it up. Probably doesn't hurt to have an extra defer Close, although if Close is transitive, that means calling it twice and the interface says that "The behavior of Close after the first call is undefined."

Reply:
The reference to the original Body is lost, so I'm not sure there's a real risk of closing it twice. On the contrary, I think none else could close it if we don't close it here.